src/v/cloud_storage/remote_partition.cc

```diff
@@ -250,16 +251,30 @@ class partition_record_batch_reader_impl final
         remote_partition::iterator iter;
     };
 
+    /// Find the starting segment for a reader
+    remote_partition::iterator
+    seek_segment(const storage::log_reader_config& config) {
+        if (config.first_timestamp) {
+            // Seek by timestamp
+            return _partition->seek_by_timestamp(*config.first_timestamp);
+        } else {
+            // Seek by offset
+            auto it = _partition->upper_bound(
+              model::offset_cast(config.start_offset));
+            if (it != _partition->begin()) {
+                it = std::prev(it);
+            }
+            return it;
+        }
+    }
+
     std::optional<cache_reader_lookup_result>
     find_cached_reader(const storage::log_reader_config& config) {
         if (!_partition || _partition->_segments.empty()) {
             return std::nullopt;
         }
-        auto it = _partition->upper_bound(
-          model::offset_cast(config.start_offset));
-        if (it != _partition->begin()) {
-            it = std::prev(it);
-        }
+
+        auto it = seek_segment(config);
```

Comment:
I think this is a better example of the pre-condition issue that wasn't clear. the pre-condition here is that manifest is non-empty, and we have a segments.empty() check above, but paths leading up to `find_cached_reader` don't obviously imply _manifest is non-empty unless of course a non-empty _segments does imply that.